toolchain/base/kind_switch.h

```diff
@@ -39,29 +39,104 @@
 // requirements should change.
 namespace Carbon::Internal::Kind {
 
+template <typename T>
+static constexpr bool IsStdVariantValue = false;
+
+template <typename... Ts>
+static constexpr bool IsStdVariantValue<std::variant<Ts...>> = true;
+
+template <typename T>
+concept IsStdVariant = IsStdVariantValue<std::decay_t<T>>;
+
+template <typename... Ts>
+struct TypePack {};
+
+template <typename T>
+struct StdVariantTypePackValue;
+
+template <typename... Ts>
+struct StdVariantTypePackValue<std::variant<Ts...>> {
+  using Types = TypePack<Ts...>;
+};
+
+template <typename T>
+using StdVariantTypePack = StdVariantTypePackValue<std::decay_t<T>>::Types;
+
+template <typename T>
+struct StdVariantEnumValue;
+
+template <typename... Ts>
+struct StdVariantEnumValue<std::variant<Ts...>> {
+  enum [[clang::enum_extensibility(open)]] EnumType{
+      VariantTypeNotHandledInSwitch = sizeof...(Ts) - 1};
+  using Type = EnumType;
+};
```

Comment:
After some testing, it looks like this trick doesn't actually work: Clang diagnoses if the *last* element of the enum doesn't have a case, but doesn't warn if lower values are not handled.

I think if we want to make this work we'll need to declare a bunch of enum types, one for each number of variant elements we want to support, up to some fixed bound. Eg:
```suggestion
template <typename T0, typename T1>
struct StdVariantEnumValue<std::variant<T0, T1>> {
  enum [[clang::enum_extensibility(open)]] EnumType{
      VariantType0,
      VariantType1,
  };
  using Type = EnumType;
};
template <typename T0, typename T1, typename T2>
struct StdVariantEnumValue<std::variant<T0, T1, T2>> {
  enum [[clang::enum_extensibility(open)]] EnumType{
      VariantType0,
      VariantType1,
      VariantType2,
  };
  using Type = EnumType;
};
template <typename T0, typename T1, typename T2, typename T3>
struct StdVariantEnumValue<std::variant<T0, T1, T2, T3>> {
  enum [[clang::enum_extensibility(open)]] EnumType{
      VariantType0,
      VariantType1,
      VariantType2,
      VariantType3,
  };
  using Type = EnumType;
};
// ...
```

Reply:
Done, with a new CARBON_FOR_EACH macro so that we can stamp out up to 12 types without it being too wild.

Also, I described all three error conditions and what they print in the PR description now.